backend/estimate-service/src/main/java/org/egov/service/NotificationService.java

```diff
@@ -0,0 +1,262 @@
+import org.apache.commons.lang3.StringUtils;
+import org.egov.common.contract.request.RequestInfo;
+import org.egov.config.EstimateServiceConfiguration;
+import org.egov.producer.Producer;
+import org.egov.repository.ServiceRequestRepository;
+import org.egov.tracer.model.CustomException;
+import org.egov.util.EstimateServiceConstant;
+import org.egov.util.HRMSUtils;
+import org.egov.util.LocationServiceUtil;
+import org.egov.util.ProjectUtil;
+import org.egov.web.models.Estimate;
+import org.egov.web.models.EstimateRequest;
+import org.egov.web.models.Workflow;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.egov.util.EstimateServiceConstant.*;
+
+@Service
+@Slf4j
+public class NotificationService {
+
+    @Autowired
+    private Producer producer;
+
+    @Autowired
+    private ServiceRequestRepository repository;
+
+    @Autowired
+    private RestTemplate restTemplate;
+
+    @Autowired
+    private EstimateServiceConfiguration config;
+
+    @Autowired
+    private HRMSUtils hrmsUtils;
+
+    @Autowired
+    private ProjectUtil projectServiceUtil;
+
+    @Autowired
+    private LocationServiceUtil locationServiceUtil;
+
+
+    /**
+     * Sends notification by putting the sms content onto the core-sms topic
+     *
+     * @param request
+     */
+    public void sendNotification(EstimateRequest request) {
+        Workflow workflow = request.getWorkflow();
+
+        if ("REJECT".equalsIgnoreCase(workflow.getAction())) {
+            pushNotificationToCreatorForRejectAction(request);
+        } else if ("APPROVE".equalsIgnoreCase(workflow.getAction())) {
+            pushNotificationToCreatorForApproveAction(request);
+        }
+    }
+
+    private void pushNotificationToCreatorForRejectAction(EstimateRequest request) {
+        Estimate estimate = request.getEstimate();
+        String createdByUuid = request.getEstimate().getAuditDetails().getCreatedBy();
+
+        log.info("get message template for reject action");
+        String message = getMessage(request);
+
+        if (StringUtils.isEmpty(message)) {
+            log.info("SMS content has not been configured for this case");
+            return;
+        }
+
+        //get project number, location, userDetails
+        log.info("get project number, location, userDetails");
+        Map<String, String> smsDetails = getDetailsForSMS(request, createdByUuid);
+
+        log.info("build Message For Reject Action");
+        message = buildMessageForRejectAction(estimate, smsDetails, message);
+        SMSRequest smsRequest = SMSRequest.builder().mobileNumber(smsDetails.get("mobileNumber")).message(message).build();
+
+        log.info("push message for REJECT Action");
+        producer.push(config.getSmsNotifTopic(), smsRequest);
+    }
+
```

Comment:
pushNotificationToCreatorForRejectAction and pushNotificationToCreatorForApproveAction method is having repetitive code. Can not be merged into single method ?

Reply:
Keeping it generic for now to accomodate for message template changes in create and update later as suggested by CK. Or else we will again need to change the code structure